Add ARC_MPY_OPTION to kconfig

YuguoWH opened the pull request:
The CONFIG_ARC_HAS_ACCL_REGS is not set for most of ARC targets. When it is set, it provides saving & restoring to r58 and r59 which are acch and accl registers. The condition to set CONFIG_ARC_HAS_ACCL_REGS should be FPU and/or MPY > 6 but there's no corresponding kconfig item for MPY. 
The issue mentioned in #39392 was caused by this, CONFIG_ARC_HAS_ACCL_REGS was not set but arcmwdt compiler uses accl and acch as general purpose registers which are corrupted during IRQ. 
This PR add a new ARC_MPY_OPTION to kconfig to patch this problem.

Fixes #39392

YuguoWH commented:
@ruuddw I think defining HAS_ACCL statically is not flexable for all arc targets. The help message for HAS_ACCL already states that it should be enabled when FPU and/or MPY > 6 but MPY is nowhere to be found. Defining MPY_OPTION has the same effect as define HAS_ACCL everywhere, but it can also provide more: We could add compiler options later so it could improve performance.

evgeniy-paltsev commented on soc/arc/snps_nsim/CMakeLists.txt:
```diff
@@ -6,48 +6,54 @@ if(COMPILER STREQUAL gcc)
+      zephyr_compile_options_ifdef(CONFIG_ARC_MPY_OPTION -mmpy-option=${CONFIG_ARC_MPY_OPTION})
     else()
       zephyr_compile_options_ifdef(CONFIG_FPU -mfpu=fpuda_all)
+      # when use GNU compiler, higher mpy-option is not supported
+      if(CONFIG_ARC_MPY_OPTION GREATER 6)
+        zephyr_compile_options(-mmpy-option=6)
+      else()
+      	zephyr_compile_options_ifdef(CONFIG_ARC_MPY_OPTION -mmpy-option=${CONFIG_ARC_MPY_OPTION})
+      endif()
```
Even if we actually need this I'd prefer to have this logic in some Cmake function/macro in arc `.cmake` and call it from SoC `CMakeLists.txt`. We can call it `arc_finalize_mpy_compile_options` and put option for both GNU and MWDT toolchain in it.

ruuddw replied:
not sure - if we want a clean, generic solution, then I come back to my earlier suggestion: it would be better to have the option to use tcf files for all hardware specific compiler options/flags.

YuguoWH commented on soc/arc/snps_nsim/CMakeLists.txt:
```diff
@@ -6,48 +6,50 @@ if(COMPILER STREQUAL gcc)
     else()
       zephyr_compile_options_ifdef(CONFIG_FPU -mfpu=fpuda_all)
+      # when use GNU compiler, higher mpy-option is not supported
+    	zephyr_compile_options_ifdef(CONFIG_ARC_MPY_OPTION -mmpy-option=2)
```
I get the confirmation from GNU team that at the moment the EM mpy value is limited to 6 as DSP exensions for EM are not supported.

abrodkin commented:
@ruuddw so then maybe we just set `CONFIG_ARC_HAS_ACCL_REGS` explicitly for all boards where those registers exists and decide which way to do its detection automatically (based on say .tcf or in any other way) later? That way we'll solve immediately that problem with corrupted registers.

ruuddw commented:
> @ruuddw so then maybe we just set `CONFIG_ARC_HAS_ACCL_REGS` explicitly for all boards where those registers exists and decide which way to do its detection automatically (based on say .tcf or in any other way) later? That way we'll solve immediately that problem with corrupted registers.

Yes, that's what I proposed as my first feedback to this PR already: "For now, why not simply define 'HAS_ACCL' for the relevant boards instead of introducing the MPY option?"

YuguoWH committed:
soc: arc: fix ARC_HAS_ACCL_REGS settings

ARC_HAS_ACCL_REGS should set to y to protect ACCL and ACCH registers
during irq. These registers could be used as GPRs by compilers and
therefore need store/restore during irq.

Signed-off-by: Yuguo Zou <yuguo.zou@synopsys.com>

YuguoWH commented:
Sorry everyone, I accidently force push the wrong branch here which triggers review requests automatically on a large scale.

YuguoWH commented:
I've run twister test with jenkins and the result proves the fix.

ruuddw reviewed (approved):
Thanks, this is much simpler and specific.

evgeniy-paltsev commented on arch/arc/Kconfig:
```diff
@@ -263,7 +263,7 @@ config CODE_DENSITY
 	bool "Reg Pair ACCL:ACCH (FPU and/or MPY > 6)"
 	default y if CPU_HS3X
-	default y if FPU
+	default y if CPU_HAS_FPU
```
BTW, the default FPU usage is without FPU_SHARING - we assume that we use FPU only in one thread, so we don't need ARC_HAS_ACCL_REGS if we have FPU by default (at least on ARC HS).

I guess we are enabling ARC_HAS_ACCL_REGS because we assuming that ACCL_REGS would be used by MPY instructions. Probably we need to add comment here that if we don't need to enable it if our system doesn't support MPY instructions / we are 100% sure that we don't use them.

ruuddw replied:
Triggered by this, I checked and actually believe FPU doesn't always mean the accl registers are there. Instead, they are only available if fused multiply & add/sub is configured in the hardware.
I think this default line can simply be removed and left to SoC configuration.

ruuddw reviewed (changes requested):
See comments, I don't believe FPU implies by default that accumulator registers are present.

ruuddw reviewed (approved):
Simple is often better, fixing the 'CONFIG_HAS_ACCL_REGS was not set' can be as simple as setting it for the targets that need this.
